# **DEMO 2**

Dataset: [Pokemon with stats (link)](https://www.kaggle.com/datasets/abcsds/pokemon)

Selection Rationale: I choose this dataset because I was recently playing the new nintendo switch game

___

### Analysis Objectives & Plan
1. You want to become a strong pokemon gym leader using the stats file! 
- First reset the index to the 'Name' of pokemon within the file.
- Condense the number of columns to visualize the pokemon type and total basepoints ('Total')
- Remove any unnecessary columns in the dataset.
2. Most gym leaders have a team make up of a particular pokemon type. Sort your list by pokemon type. 
- Sort the pokemon with the highest HP (health points) and type.
- Sort pokemon with highest overall (Total) points & create a list based on strongest pokemon in a descedning order. 
- Now choose a type based on the top few strongest. Organize the list to have your chosen type and highest Total points.
3. The strongest pokemon within certain types may not be possible to capture or aviable in all games. Generate the list of realistic pokemon you can use. 
- Remove options you likely won't be able to obtain within the game (i.e. legendary)
- Provide descriptive statistics for your new list. 
- What is the mean values of the new list
- Based on highest speed stats choose your team of 6 pokemon

Love the enthusiasm for the first objective, but it's a bit unclear what the analysis objective is - what exactly are we looking to find? (e.g., You could write, "Look at pokemon types and see base points). Otherwise - the rest looks great here!

___

### Import Libraries 

In [1]:
# import necessary libraries 
import pandas as pd 

___

## Import and Explore Data
- import dataset and assign to a variable named df
- print/output the following (as pythonically as possible):
    - DataFrame's shape
    - Column names 
    - First 5 rows 

In [2]:
# import data from csv file
df = pd.read_csv('Pokemon2.csv')

In [3]:
df 

#                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander     Fire     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion     Fire   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
3        122      120     80           1      False  
4         60       50     65           1      False  
..       ...      ...    ...         ...        ...  
795      100      150     50           6       True  
796      160      110    110           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[800 rows x 13 columns]

Always good practice to include a comment indicating the purpose of the code.

In [4]:
# df shape
print(df.shape)

(800, 13)


In [5]:
# df columns 
print(df.columns)

Index(['#', 'Name', 'Type 1', 'Type 2', 'Total', 'HP', 'Attack', 'Defense',
       'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')


In [7]:
# output first 5 rows 
print(df.head())

   #                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0  1              Bulbasaur  Grass  Poison    318  45      49       49   
1  2                Ivysaur  Grass  Poison    405  60      62       63   
2  3               Venusaur  Grass  Poison    525  80      82       83   
3  3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4  4             Charmander   Fire     NaN    309  39      52       43   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0       65       65     45           1      False  
1       80       80     60           1      False  
2      100      100     80           1      False  
3      122      120     80           1      False  
4       60       50     65           1      False  


___

Nice job exploring the data!

### Import and Explain Tasks 

___

##### Objective 1

In [8]:
# Step 1. We don't really care about the pokemon number to reassign Name to be the first column
df = df.set_index('Name')


In [9]:
df

#   Type 1  Type 2  Total  HP  Attack  Defense  \
Name                                                                      
Bulbasaur                1    Grass  Poison    318  45      49       49   
Ivysaur                  2    Grass  Poison    405  60      62       63   
Venusaur                 3    Grass  Poison    525  80      82       83   
VenusaurMega Venusaur    3    Grass  Poison    625  80     100      123   
Charmander               4     Fire     NaN    309  39      52       43   
...                    ...      ...     ...    ...  ..     ...      ...   
Diancie                719     Rock   Fairy    600  50     100      150   
DiancieMega Diancie    719     Rock   Fairy    700  50     160      110   
HoopaHoopa Confined    720  Psychic   Ghost    600  80     110       60   
HoopaHoopa Unbound     720  Psychic    Dark    680  80     160       60   
Volcanion              721     Fire   Water    600  80     110      120   

                       Sp. Atk  Sp. Def  Speed  Generation  Legendary  
Name                                                                   
Bulbasaur                   65       65     45           1      False  
Ivysaur                     80       80     60           1      False  
Venusaur                   100      100     80           1      False  
VenusaurMega Venusaur      122      120     80           1      False  
Charmander                  60       50     65           1      False  
...                        ...      ...    ...         ...        ...  
Diancie                    100      150     50           6       True  
DiancieMega Diancie        160      110    110           6       True  
HoopaHoopa Confined        150      130     70           6       True  
HoopaHoopa Unbound         170      130     80           6       True  
Volcanion                  130       90     70           6       True  

[800 rows x 12 columns]

Should include a comment in the code cell above to indicate its purpose.

In [10]:
#Step 2. We wish to see primarily the pokemon type and total base points. Index the table to show these in a simplier view
df.iloc[:-1, 0:4] 

#   Type 1  Type 2  Total
Name                                              
Bulbasaur                1    Grass  Poison    318
Ivysaur                  2    Grass  Poison    405
Venusaur                 3    Grass  Poison    525
VenusaurMega Venusaur    3    Grass  Poison    625
Charmander               4     Fire     NaN    309
...                    ...      ...     ...    ...
Zygarde50% Forme       718   Dragon  Ground    600
Diancie                719     Rock   Fairy    600
DiancieMega Diancie    719     Rock   Fairy    700
HoopaHoopa Confined    720  Psychic   Ghost    600
HoopaHoopa Unbound     720  Psychic    Dark    680

[799 rows x 4 columns]

In [11]:
#Step 3. We don't care which generation the pokemon is from, so we can drop this column
df = df.drop(columns = 'Generation')

Code looks good here!

___

##### Objective 2

In [12]:
# step 1. We want to see the pokemon with the highest HP sorted by Type. Sort the table to show this
df_sorted = df.sort_values(['Type 1', 'HP'], ascending=[1,0])

In [13]:
print(df_sorted)

                           # Type 1    Type 2  Total  HP  Attack  Defense  \
Name                                                                        
Yanmega                  469    Bug    Flying    515  86      76       86   
Volcarona                637    Bug      Fire    550  85      60       65   
Heracross                214    Bug  Fighting    500  80     125       75   
HeracrossMega Heracross  214    Bug  Fighting    600  80     185      115   
Accelgor                 617    Bug       NaN    495  80      70       40   
...                      ...    ...       ...    ...  ..     ...      ...   
Krabby                    98  Water       NaN    325  30     105       90   
Horsea                   116  Water       NaN    295  30      40       70   
Staryu                   120  Water       NaN    340  30      45       55   
Magikarp                 129  Water       NaN    200  20      10       55   
Feebas                   349  Water       NaN    200  20      15       20   

The last two code cells could be combined into one cell.

In [14]:
#Step 2. We want pokemon with high overall (Total) points, create a list based on strongest pokemon in a descedning order. 
condition = df['Total'] > 500
sorted_df = df[condition].sort_values(by='Total', ascending=False)
print(sorted_df)

                         #    Type 1    Type 2  Total   HP  Attack  Defense  \
Name                                                                          
RayquazaMega Rayquaza  384    Dragon    Flying    780  105     180      100   
MewtwoMega Mewtwo X    150   Psychic  Fighting    780  106     190      100   
MewtwoMega Mewtwo Y    150   Psychic       NaN    780  106     150       70   
GroudonPrimal Groudon  383    Ground      Fire    770  100     180      160   
KyogrePrimal Kyogre    382     Water       NaN    770  100     150       90   
...                    ...       ...       ...    ...  ...     ...      ...   
Machamp                 68  Fighting       NaN    505   90     130       80   
Ninetales               38      Fire       NaN    505   73      76       75   
Nidoking                34    Poison    Ground    505   81     102       77   
Nidoqueen               31    Poison    Ground    505   90      92       87   
Conkeldurr             534  Fighting       NaN    50

___

In [15]:
#Step 3. You decided to choose psychic pokemon as your type to train. Organize the list to have psychic type and highest Total points.
psychic_condition = df['Type 1'] == 'Psychic'
hp_condition = df['Total'] > 500
combined_condition = psychic_condition & hp_condition
psychic_pokemon_high_hp = df[combined_condition].sort_values(by='Total', ascending=False)
print(psychic_pokemon_high_hp)

                           #   Type 1    Type 2  Total   HP  Attack  Defense  \
Name                                                                           
MewtwoMega Mewtwo X      150  Psychic  Fighting    780  106     190      100   
MewtwoMega Mewtwo Y      150  Psychic       NaN    780  106     150       70   
HoopaHoopa Unbound       720  Psychic      Dark    680   80     160       60   
Lugia                    249  Psychic    Flying    680  106      90      130   
Mewtwo                   150  Psychic       NaN    680  106     110       90   
GalladeMega Gallade      475  Psychic  Fighting    618   68     165       95   
GardevoirMega Gardevoir  282  Psychic     Fairy    618   68      85       65   
DeoxysDefense Forme      386  Psychic       NaN    600   50      70      160   
HoopaHoopa Confined      720  Psychic     Ghost    600   80     110       60   
Victini                  494  Psychic      Fire    600  100     100      100   
Cresselia                488  Psychic   

Output looks good in this objective section!


##### Objective 3

In [16]:
# step 1. Realistically, you will not have legendary pokemon on your team. Filter your new list to remove legendary pokemon.
new_table_rows = df[(df['Type 1'] == 'Psychic') & (df['Total'] > 500) & (df['Legendary'] == False)]
print(new_table_rows)

                           #   Type 1    Type 2  Total   HP  Attack  Defense  \
Name                                                                           
AlakazamMega Alakazam     65  Psychic       NaN    590   55      50       65   
Mew                      151  Psychic       NaN    600  100     100      100   
Espeon                   196  Psychic       NaN    525   65      65       60   
Celebi                   251  Psychic     Grass    600  100     100      100   
Gardevoir                282  Psychic     Fairy    518   68      65       65   
GardevoirMega Gardevoir  282  Psychic     Fairy    618   68      85       65   
Gallade                  475  Psychic  Fighting    518   68     125       65   
GalladeMega Gallade      475  Psychic  Fighting    618   68     165       95   
Cresselia                488  Psychic       NaN    600  120      70      120   

                         Sp. Atk  Sp. Def  Speed  Legendary  
Name                                                     

In [17]:
# step 2. What are the descriptive statistics of your new list?
df = new_table_rows
df.describe()

#       Total          HP      Attack     Defense     Sp. Atk  \
count    9.000000    9.000000    9.000000    9.000000    9.000000    9.000000   
mean   296.111111  576.333333   79.111111   91.666667   81.666667  111.111111   
std    153.194684   42.982555   21.848595   35.881750   22.079402   40.833333   
min     65.000000  518.000000   55.000000   50.000000   60.000000   65.000000   
25%    196.000000  525.000000   68.000000   65.000000   65.000000   75.000000   
50%    282.000000  600.000000   68.000000   85.000000   65.000000  100.000000   
75%    475.000000  600.000000  100.000000  100.000000  100.000000  130.000000   
max    488.000000  618.000000  120.000000  165.000000  120.000000  175.000000   

          Sp. Def       Speed  
count    9.000000    9.000000  
mean   111.111111  101.666667  
std     14.743172   21.505813  
min     95.000000   80.000000  
25%    100.000000   85.000000  
50%    115.000000  100.000000  
75%    115.000000  110.000000  
max    135.000000  150.000000

In [19]:
# step 3. What is the mean of your new list?
df.mean()

TypeError: unsupported operand type(s) for +: 'int' and 'str'

This code threw an error. Also, this code cell isn't necessary because you can already view the means in the previous output.

In [20]:
#Step 4. Based on highest speed stats choose your team of 6 pokemon.
my_team = ['AlakazamMega Alakazam', 'Espeon', 'Celebi', 'Mew', 'GardevoirMega Gardevoir', 'GalladeMega Gallade']

In [21]:
df.loc[my_team]

#   Type 1    Type 2  Total   HP  Attack  Defense  \
Name                                                                           
AlakazamMega Alakazam     65  Psychic       NaN    590   55      50       65   
Espeon                   196  Psychic       NaN    525   65      65       60   
Celebi                   251  Psychic     Grass    600  100     100      100   
Mew                      151  Psychic       NaN    600  100     100      100   
GardevoirMega Gardevoir  282  Psychic     Fairy    618   68      85       65   
GalladeMega Gallade      475  Psychic  Fighting    618   68     165       95   

                         Sp. Atk  Sp. Def  Speed  Legendary  
Name                                                         
AlakazamMega Alakazam        175       95    150      False  
Espeon                       130       95    110      False  
Celebi                       100      100    100      False  
Mew                          100      100    100      False  
GardevoirMega Gardevoir      165      135    100      False  
GalladeMega Gallade           65      115    110      False

Output looks good! Just would want a comment indicating what the code is doing.

___

##### Conclusions

Based on filtering for total stat points, removing uncessasy columns within the dataset, sorting by type, and sorting by ascending statistics we were able to generate a list based on strong types & unltimating deciding on a psychic type list. Additional conditions based on legendary status allowed us to generate a team of 6 would be realisitically capture & train within a game to run a psychic style gym. 

We narrowed down our list in the end & could make our final choice based on another stat (speed) since the list was condensed. We could then generate basic descriptive stats of this list with the MEAN showing the most useful information. This indicated our teams mean total base stats was 576.333333, which is good setting our threshold at >500 and the highest in the game being 780 (for a legendary). 

You would want to remove the markdown cell above (you wouldn't want extraneous cells). Overall, nice work explaining what sort of information your analysis can tell us!